k8s/upgrade/src/preflight_validations.rs

```
@@ -45,6 +54,71 @@ pub async fn preflight_check(
+    let source_version = get_source_version(ns).await?;
+    let destination_version = get_image_version_tag();
+    if source_version.eq(&destination_version) {
+        console_logger::error(
+            user_prompt::SOURCE_TARGET_VERSION_SAME,
+            &destination_version,
+        );
+        return Err(Error::SourceTargetVersionSame);
+    }
+
+    let valid_paths = parse_key_value_pairs(VALID_UPGRADE_PATHS);
+    let data = get_valid_upgrade_paths(valid_paths.clone());
+    match valid_paths.get(&source_version) {
+        Some(target_set) => {
+            if !target_set.contains(&destination_version) {
+                console_logger::error(user_prompt::NOT_A_VALID_UPGRADE_PATH, data.as_str());
+                return Err(Error::NotAValidUpgradePath);
+            }
+        }
+        None => {
+            console_logger::error(user_prompt::NOT_A_VALID_SOURCE_FOR_UPGRADE, data.as_ref());
+            return Err(Error::NotAValidSourceForUpgrade);
+        }
+    }
+    Ok(())
+}
+
+// This function creates the output valid upgrade paths from the map.
+fn get_valid_upgrade_paths(upgrade_paths: HashMap<String, HashSet<String>>) -> String {
+    let mut valid_paths: String = String::new();
+    for (key, val) in upgrade_paths.iter() {
+        for dest in val.iter() {
+            valid_paths.push_str(key);
+            valid_paths.push_str(" -> ");
+            valid_paths.push_str(dest);
+            valid_paths.push('\n');
+        }
+    }
+    valid_paths
+}
+
+// The function parses VALID_UPGRADE_PATHS and creates map
+// where key is source version value is hashset of destination versions.
+fn parse_key_value_pairs(input: &str) -> HashMap<String, HashSet<String>> {
+    input
+        .split(';')
+        .filter(|s| !s.is_empty())
+        .map(|s| {
+            let mut iter = s.split(':');
+            (
+                iter.next().unwrap_or("").to_string(),
+                iter.next().unwrap_or("").to_string(),
```

Is "" the right approach or should we return an error here?